### PR TITLE
feat: get surveys api

### PR DESCRIPTION
**Pull request by liyiy**
## Changes

Add `getSurveys` call for JS library
...

## Checklist
- [ ] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers

### CODE REVIEW

File: src/posthog-surveys.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { PostHog } from "./posthog-core";
+import { SURVEYS } from "./posthog-persistence";
+import { SurveyCallback } from "types";
+
+
+export class PostHogSurveys {
```
Comment (neilkakkar) [inline src/posthog-surveys.ts:6]: I'm assuming the other necessary API targeting options will get added to this class in a follow up?

**Review by neilkakkar — APPROVED** [review]

File: src/__tests__/surveys.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            { name: 'first survey', description: 'first survey description', type: SurveyType.Popover, questions: [{ type: SurveyQuestionType.Open, question: 'what is a bokoblin?' }] },
+        ]
+
+    const secondSurveys = [
+        { name: 'first survey', description: 'first survey description', type: SurveyType.Popover, questions: [{ type: SurveyQuestionType.Open, question: 'what is a bokoblin?' }] },
+        { name: 'second survey', description: 'second survey description', type: SurveyType.Popover, questions: [{ type: SurveyQuestionType.Open, question: 'what is a moblin?' }] },
+    ]
+    // possibly compute whether the survey is active or not for the user since we already have feature flags available here.. with linked/targeting flags
+
+    given('surveysResponse', () => ({ surveys: firstSurveysResponse }))
```
Comment (neilkakkar) [inline src/__tests__/surveys.js:41]: I can't find definition for this

File: src/__tests__/surveys.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        )
+        expect(given.instance._send_request).toHaveBeenCalledTimes(1)
+        expect(given.instance.persistence.props.$surveys).toEqual(firstSurveys)
+
+        given('surveysResponse', () => ({ surveys: secondSurveys }))
+
+        given.surveys.getSurveys((data) => {
+            expect(data).toEqual(secondSurveysResponse)
+        }, true)
+        expect(given.instance._send_request).toHaveBeenCalledTimes(2)
```
Comment (neilkakkar) [inline src/__tests__/surveys.js:82]:
```suggestion
        expect(given.instance._send_request).toHaveBeenCalledTimes(2)
        expect(given.instance.persistence.props.$surveys).toEqual(secondSurveys)
```

File: src/posthog-surveys.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    instance: PostHog
+
+    constructor(instance: PostHog) {
+        this.instance = instance
+    }
+
+    getSurveys(callback: SurveyCallback, forceReload: boolean = false) {
+        const existingSurveys = this.instance.get_property('$surveys')
```
Comment (neilkakkar) [inline src/posthog-surveys.ts:14]:
```suggestion
        const existingSurveys = this.instance.get_property(SURVEYS)
```

File: src/posthog-surveys.ts (hunk 1 above)
Comment (liyiy) [reply src/posthog-surveys.ts:6]: Are you talking about for the linked/targeted feature flags?

File: src/posthog-surveys.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                    this.instance.persistence.register({ [SURVEYS]: surveys })
+                    return callback(surveys)
+                }
+            )
+        } else {
+            return callback(existingSurveys)
+        }
+    }
+
+    getActiveMatchingSurveys(callback: SurveyCallback, forceReload = false) {
```
Comment (liyiy) [inline src/posthog-surveys.ts:30]: @neilkakkar I added this and tests for it

File: src/posthog-surveys.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        this.getSurveys((surveys) => {
+            const activeSurveys = surveys.filter((survey) => {
+                return !!(survey.start_date && !survey.end_date)
```
Comment (neilkakkar) [inline src/posthog-surveys.ts:33]: we ought to check this date in relation to `now` as well, i.e. start < now < end. I'm pretty sure we'd want surveys to be schedulable, which is where this check comes in handy

File: src/posthog-surveys.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            })
+            const conditionMatchedSurveys = activeSurveys.filter((survey) => {
+                if (!survey.conditions) {
+                    return true
+                }
+                if (survey.conditions?.url && !survey.conditions.selector) {
```
Comment (neilkakkar) [inline src/posthog-surveys.ts:39]: suggestion: You can simplify these checks like so:

```
const urlCheck = survey.conditions?.url ? window.location.href.indexOf(survey.conditions.url) > -1 : true
```

and then just `return urlCheck && selectorCheck`.

Same for below.

**Review by neilkakkar — APPROVED** [review]

File: src/posthog-surveys.ts (hunk 6 above)
Comment (liyiy) [reply src/posthog-surveys.ts:33]: yeah though we don't need it now until we have scheduling and such, which will require an update to the logic and tests then, wyt?

File: src/posthog-surveys.ts (hunk 7 above)
Comment (liyiy) [reply src/posthog-surveys.ts:39]: yass

File: src/posthog-surveys.ts (hunk 6 above)
Comment (neilkakkar) [reply src/posthog-surveys.ts:33]: yep all good

**Pull request merged**